### PR TITLE
Sentry: Add exception event capture in sentry

### PR DESCRIPTION
**Pull request by uds5501**
Signed-off-by: Uddeshya Singh <singhuddeshyaofficial@gmail.com>

**Description:** <Describe what has changed.>
This commit will iterate through span events and convert the `exception` events to their equivalent sentry error events.

This handling has been inspired by `elastic-exporter` encoding of events enclosed here : https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9c7c329f0877f72befb36c61c47899d5a4525037/exporter/elasticexporter/internal/translator/elastic/traces.go#L376-L402 

**Link to tracking Issue:**  #3809 

**Testing:** None yet

**Documentation:** None yet

### CODE REVIEW

**Comment by uds5501** [comment]
@AbhiPrasad Please take a look :D

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter.go:80]: I would prefer if the mutated arg is the first one passed in, makes it more apparent that the function’s side effect is changing a passed in arg.

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter.go:60]: We can’t estimate the number of span events in anyway? I would encourage atleast setting the capacity to 10 so we minimize allocations.

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter.go:112]: Well this is no longer only transactions. Can we rename to events? We’ll have to rename the transport method name was well, `CaptureEvent`?

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter.go:136]: Can we indicate that `eventList` is mutated here in the func docstring?

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter.go:182]: Why are we appending? New events should have no existing exception right?

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter.go:176]: You’ll have to use the `statusFromSpanStatus` function defined below to convert span status to a sentry status.

Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:176]: We could maybe even only compute the status once.

**Review by AbhiPrasad — CHANGES_REQUESTED** [review]
Great work, thanks for your contribution!

Aside from addressing the review comments, let’s make sure we write some tests for this new functionality.

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:60]: Hmm actually maybe 10 is too much, thoughts on what we should do here?

**Comment by uds5501** [comment]
Thanks for the review @AbhiPrasad , I will take a look at it in a bit.

File: exporter/sentryexporter/sentry_exporter.go
Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter.go:80]: Will address this in next commit.

File: exporter/sentryexporter/sentry_exporter.go
Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter.go:112]: events sounds good, as far as the transport function is concerned, how about `SendEvents`?

File: exporter/sentryexporter/sentry_exporter.go
Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter.go:182]: Fair point, will be fixing this in the next commit.

File: exporter/sentryexporter/sentry_exporter.go
Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter.go:176]: Actually, over here the `span` is not `pdata.Span`, it's actually `*sentry.Span`, so there is no conversion required here.

File: exporter/sentryexporter/sentry_exporter.go
Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter.go:60]: We can start by setting the capacity to 5/10, because currently I don't see any method to pre-calculate number of recorded exception events, and if we do write one such helper method, it'd be an unnecessary computation overhead because of going through all the span events twice.

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:176]: Whoops lol, sorry about that

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:112]: Yeah `SendEvents` works for me

File: exporter/sentryexporter/sentry_exporter.go
Comment (rhcarvalho) [reply exporter/sentryexporter/sentry_exporter.go:60]: The Go runtime takes care of deciding how to allocate the backing array, we should not need to over-optimize at this point.

In fact, it is idiomatic to not allocate slices with `make` if it may end up empty (thus, no allocation).

```go
var exceptionEvents []*sentry.Event
...
exceptionEvents = append(exceptionEvents, event)  // dynamically allocated, when needed
```

Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter.go]: @rhcarvalho , acknowledged, in this commit I have adopted your suggestions in the current commit.

**Comment by uds5501** [comment]
@AbhiPrasad I put up a new commit regarding last review's resolutions.
Also, I think I need help for the tests.

I tried running :

```cmd
$ go test sentry_exporter_test.go
```

but the terminal threw me this,
```cmd
./sentry_exporter_test.go:49:14: undefined: unixNanoToTime
./sentry_exporter_test.go:50:14: undefined: unixNanoToTime
./sentry_exporter_test.go:65:14: undefined: unixNanoToTime
./sentry_exporter_test.go:66:14: undefined: unixNanoToTime
./sentry_exporter_test.go:81:14: undefined: unixNanoToTime
./sentry_exporter_test.go:82:14: undefined: unixNanoToTime
./sentry_exporter_test.go:95:14: undefined: unixNanoToTime
./sentry_exporter_test.go:96:14: undefined: unixNanoToTime
./sentry_exporter_test.go:106:17: undefined: unixNanoToTime
./sentry_exporter_test.go:107:17: undefined: unixNanoToTime
./sentry_exporter_test.go:107:17: too many errors
FAIL    command-line-arguments [build failed]
FAIL
```

So, is this the correct way to run tests? Or am I missing something? I need this to write my own tests tbh.

**Comment by rhcarvalho** [comment]
> So, is this the correct way to run tests? Or am I missing something?

From the root of the repository, run:

```
go test ./exporter/sentryexporter/
```

**Comment by uds5501** [comment]
> > So, is this the correct way to run tests? Or am I missing something?
> 
> From the root of the repository, run:
> 
> ```
> go test ./exporter/sentryexporter/
> ```
```
main module (github.com/open-telemetry/opentelemetry-collector-contrib) does not contain package github.com/open-telemetry/opentelemetry-collector-contrib/exporter/sentryexporter
```
What should I do? :(

**Comment by rhcarvalho** [comment]
Ah, I'm sorry, typing from a phone before and now. That message is because the exporters are on separate modules.

You can try:

```
cd ./exporter/sentryexporter
go test
```

File: exporter/sentryexporter/sentry_exporter.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter.go:46]: This shouldn’t be a global variable, it should be in the `pushTraceData` func.

Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter.go:46]: Okay, will change.

**Comment by uds5501** [comment]
@AbhiPrasad @rhcarvalho 
Added test and fixed some errors

File: exporter/sentryexporter/sentry_exporter_test.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter_test.go:173]: Can we add table tests where the exception type or exception message is empty? Just to say that these are valid arguments to pass into the func?

**Review by AbhiPrasad — COMMENTED** [review]
Almost there! Can we please have tests for `convertEventsToSentryExceptions()`? Ideally I'd like to see a table test that covers different combinations of exception and regular span events.

I think we are good to merge after having those tests. Thanks so much for your work here!

File: exporter/sentryexporter/sentry_exporter_test.go
Comment (rhcarvalho) [inline exporter/sentryexporter/sentry_exporter_test.go:26]: Please undo the changes to imports here. Standard library imports are conventionally grouped together as the first group in an `import (...)` statement.

Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter_test.go:26]: Okay

**Comment by uds5501** [comment]
@AbhiPrasad regarding `convertEventsToSentryExceptions()`, that function is essentially calling essentially calling `sentryEventFromError()` and adding that new event to an array. That's why I thought directly testing `sentryEventFromError()` would be okay.

As for adding exception type / message as empty, I will add that test.

**Comment by uds5501** [comment]
@AbhiPrasad @rhcarvalho New changes pushed

**Review by AbhiPrasad — COMMENTED** [review]
2 more things and I think we are ready for the maintainers to review. Appreciate your work on this!

**Comment by uds5501** [comment]
@AbhiPrasad Added the modified the testing variant.
But for some reason `make lint` wasn't working.
I ran it in the root folder (not exporter/sentryexporter) , where should I be running it to finish the linting?

**Comment by AbhiPrasad** [comment]
Running `make lint` in the `exporter/sentryexporter` folder should be enough. Make sure you run `make install-tools` in the root directory to install the tools necessary.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0e6a714f9f2993eb9b949f9ec942770daa2220c6/Makefile#L160-L165

File: exporter/sentryexporter/sentry_exporter_test.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter_test.go:250]: Haven't tested this, but hopefully you get the approach. We can use a func to simplify the logic when generating the expected event. Can do the same with the test below. Means we don't have a ton of test case specific boilerplate above the test cases. Instead the logic above the test cases are general to every test case.

```suggestion
			testName:            "Exception Event with both exception type and message",
			errorMessage:        errorMessage,
			errorType:           errorType,
			sampleSentrySpan:    sampleSentrySpanForEvent,
			expectedSentryEvent: func() {
			    expectedSentryEventWithTypeAndMessage := sentryEventBase
	                    expectedSentryEventWithTypeAndMessage.Type = errorType
	                    expectedSentryEventWithTypeAndMessage.Message = errorMessage
	                    expectedSentryEventWithTypeAndMessage.Exception = []sentry.Exception{{
		                    Value: errorMessage,
		                    Type:  errorType,
	                    }}
	                    return &expectedSentryEventWithTypeAndMessage
			}(),
			isErrorExpected:     false,
			expectedError:       nil,
```

File: exporter/sentryexporter/sentry_exporter_test.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter_test.go:253]:
```suggestion
			testName:            "Exception Event with only exception type",
```

We probably need a quick test with only message: `"Exception Event with only exception message"`

Will be easier if you use the func approach from above.

File: exporter/sentryexporter/sentry_exporter_test.go
Comment (AbhiPrasad) [inline exporter/sentryexporter/sentry_exporter_test.go:255]:
```suggestion
			errorType:           errorType,
```

Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter_test.go:255]: Acknowledged !

File: exporter/sentryexporter/sentry_exporter_test.go
Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter_test.go:253]: Understood!

File: exporter/sentryexporter/sentry_exporter_test.go
Comment (uds5501) [reply exporter/sentryexporter/sentry_exporter_test.go:250]: This is a trick I didn't think of, thanks for mentioning it!

**Comment by uds5501** [comment]
Hi @AbhiPrasad thank you so much for your follow ups and your patience. I have rewritten the tests according to your suggestions and new PR is ready for review.

**Comment by uds5501** [comment]
@AbhiPrasad @rhcarvalho what's wrong with loadtest? any guidance on that? 😢

**Comment by AbhiPrasad** [comment]
Might be a flaky test, seems to be related to the OTLP exporter. Check locally with `make e2e-test` to confirm. If it is a flake, try re-running tests in circleci, or pushing up an empty commit (`git commit --amend-empty -m "commit to rerun tests`) to restart CI.

**Comment by uds5501** [comment]
@AbhiPrasad Done

**Review by AbhiPrasad — APPROVED** [review]
I can ✅ on the Sentry side - @dmitryax this is ready for general otel collector maintainer review. Thanks for contribution @uds5501!

**Comment by uds5501** [comment]
@AbhiPrasad @dmitryax is the development in freeze or something?

**Comment by AbhiPrasad** [comment]
You might be able to get a maintainer to take a look by pinging the #opentelemetry-collector channel on Slack http://slack.cncf.io/

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
